core/state_transition.go

```diff
@@ -163,12 +165,69 @@ func IntrinsicGas(data []byte, accessList types.AccessList, isContractCreation b
 		gas += z * params.TxDataZeroGas
 	}
 	if accessList != nil {
+		accessListGas, err := accessListGas(rules, accessList)
+		if err != nil {
+			return 0, err
+		}
+		totalGas, overflow := commonMath.SafeAdd(gas, accessListGas)
+		if overflow {
+			return 0, fmt.Errorf("overflow adding access list gas")
+		}
+		gas = totalGas
+	}
+
+	return gas, nil
+}
+
+func accessListGas(rules params.Rules, accessList types.AccessList) (uint64, error) {
+	var gas uint64
+	if !rules.PredicatesExist() {
 		gas += uint64(len(accessList)) * params.TxAccessListAddressGas
 		gas += uint64(accessList.StorageKeys()) * params.TxAccessListStorageKeyGas
+		return gas, nil
 	}
+
+	for i, accessTuple := range accessList {
+		address := accessTuple.Address
+		if !rules.PredicateExists(address) {
+			// Previous access list gas calculation does not use safemath because an overflow would not be possible with
+			// the size of access lists that could be included in a block and standard access list gas costs.
+			// Therefore, we only check for overflow when adding to [totalGas], which could include the sum of values
+			// returned by a predicate.
+			accessTupleGas := params.TxAccessListAddressGas + uint64(len(accessTuple.StorageKeys))*params.TxAccessListStorageKeyGas
+			totalGas, overflow := commonMath.SafeAdd(gas, accessTupleGas)
+			if overflow {
+				return 0, fmt.Errorf("overflow adding storage slot gas at index %d", i)
+			}
+			gas = totalGas
+		} else {
+			predicateGas, err := ApplyPredicateGas(rules, accessTuple)
+			if err != nil {
+				return 0, err
+			}
+			totalGas, overflow := commonMath.SafeAdd(gas, predicateGas)
+			if overflow {
+				return 0, fmt.Errorf("overflow adding predicate gas at index %d", i)
+			}
+			gas = totalGas
+		}
+	}
+
 	return gas, nil
 }
 
+func ApplyPredicateGas(rules params.Rules, accessTuple types.AccessTuple) (uint64, error) {
+	predicate, ok := rules.PredicatePrecompiles[accessTuple.Address]
```

Comment:
does this function also need to be public?

Reply:
Nope good callout